crates/metrics/src/label.rs

```diff
@@ -0,0 +1,545 @@
+/// not just [`String`]s. The most frequent use-case (aside from the actual
+/// strings) - numbers.
+///
+/// Due to the lack of `&'static str` const generics at the moment the label
+/// name should be specified using the following hack:
+///
+/// ```
+/// use wc_metrics::{label_name, StringLabel};
+///
+/// type MyLabel = StringLabel<{ label_name("my_label") }>;
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct StringLabel<const NAME: u128, T = String>(pub T);
+
+impl<const NAME: LabelName, T> StringLabel<NAME, T> {
+    /// Creates a new [`StringLabel`].
+    ///
+    /// Expects a [`Borrow`]ed version of your type as the owned version is not
+    /// needed for the label resolution.
+    pub fn new<U: ?Sized>(ref_: &U) -> StringLabel<NAME, &U>
+    where
+        T: Borrow<U>,
+    {
+        StringLabel(ref_)
+    }
+
+    /// Converts this [`StringLabel`] into the inner `T`.
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<const NAME: LabelName, T> AsRef<T> for StringLabel<NAME, T> {
+    fn as_ref(&self) -> &T {
+        &self.0
+    }
+}
+
+pub struct StringCollection<T, M: 'static> {
+    inner: ArcSwap<HashMap<T, &'static M>>,
+    mutex: Mutex<()>,
+    attrs: Attrs,
+}
+
+impl<const NAME: LabelName, T, M> DynamicLabel<M> for StringLabel<NAME, T>
+where
+    M: 'static,
+{
+    type MetricCollection = StringCollection<T, M>;
+}
+
+impl<const NAME: LabelName, T, M> Metric for WithLabel<StringLabel<NAME, T>, M>
+where
+    M: Metric + 'static,
+{
+    fn register(attrs: &Attrs) -> Self {
+        Self {
+            collection: StringCollection {
+                inner: ArcSwap::new(Arc::new(HashMap::new())),
+                mutex: Mutex::new(()),
+                attrs: attrs.clone(),
+            },
+        }
+    }
+}
+
+impl<const NAME: LabelName, T, U, M> ResolveLabels<(StringLabel<NAME, &U>,)>
+    for WithLabel<StringLabel<NAME, T>, M>
+where
+    T: std::hash::Hash + Eq + Borrow<U> + ToString + Clone,
+    U: std::hash::Hash + Eq + ToOwned<Owned = T> + ?Sized,
+    M: Metric + 'static,
+{
+    type Target = M;
+
+    fn resolve_labels(&self, (label,): (StringLabel<NAME, &U>,)) -> &M {
+        let label = label.0;
+        let col = &self.collection;
+
+        if let Some(m) = col.inner.load().get(label) {
+            return m;
+        };
+
+        let _guard = col.mutex.lock();
+
+        let inner = col.inner.load();
+
+        // In case if another thread has already initialized the metric while we were
+        // waiting on the lock
+        if let Some(m) = inner.get(label) {
+            return m;
+        };
+
+        // Copy-on-write
+        let m: &'static M = {
+            // Make a deep copy of the `HashMap`.
+            let mut inner_clone: HashMap<_, _> = (**inner).clone();
+
+            let name = const { resolve_label_name::<NAME>() };
+            let label_ = Label::new(name, label.to_owned().to_string());
```

Comment:
What's up with the `_` suffix?

Reply:
I didn't have the brain power to come up with a different name, basically there are two labels in scope and both of them are being use later. It's similar to the math notation when you have variable `a` and `a'`

Reply:
Looks weird...